### PR TITLE
basic auth: optimize code and improve the exception/response message

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: basic auth: optimize code and improve the exception/response message
Additional Description:
Risk Level: minor.
Testing: unit.
Docs Changes: n/a.
Release Notes: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
cc @zhaohuabing could you take a first pass when you have free time? Thanks.

File: source/extensions/filters/http/basic_auth/basic_auth_filter.cc [hunk 1]
```diff
@@ -43,46 +42,49 @@ BasicAuthFilter::BasicAuthFilter(FilterConfigConstSharedPtr config) : config_(st
...
+  absl::string_view username = decoded_view.substr(0, colon_pos);
+  absl::string_view password = decoded_view.substr(colon_pos + 1);
+
+  if (!config_->validateUser(username, password)) {
+    return onDenied("User authentication failed. Invalid username/password combination.",
+                    "invalid_credential_for_basic_auth");
+  }
+
+  config_->stats().allowed_.inc();
+  return Http::FilterHeadersStatus::Continue;
```
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.cc:79]: This update treat `no auth header`, `has auth header but not for basic auth` and `has auth header but no colon` as three different error.

File: source/extensions/filters/http/basic_auth/config.cc [hunk 2]
```diff
@@ -12,36 +12,50 @@ using envoy::extensions::filters::http::basic_auth::v3::BasicAuth;
...
-        continue;
-      }
+    hash = hash.substr(5);
+    // The base64 encoded SHA1 hash is 28 bytes long
+    if (hash.length() != 28) {
+      throw EnvoyException(fmt::format("basic auth: invalid SHA hash length for '{}'", name));
     }
 
-    throw EnvoyException("basic auth: unsupported htpasswd format: please use {SHA}");
+    users.insert({name, {name, hash}});
```
Comment (wbpcode) [inline source/extensions/filters/http/basic_auth/config.cc:58]: - handle empty lines and comment lines.
- handle duplicate users.
- handle empty hash.
- tread no colon and non SHA as different format error.

File: source/extensions/filters/http/basic_auth/config.cc [hunk 3]
```diff
@@ -12,36 +12,50 @@ using envoy::extensions::filters::http::basic_auth::v3::BasicAuth;
...
-        throw EnvoyException("basic auth: invalid user name");
-      }
+    if (name.empty() || hash.empty()) {
+      throw EnvoyException("basic auth: empty user name or password");
+    }
+
+    // TODO(wbpcode): user name in exception message would simplify the debugging. But if we think
+    // it also sensitive information, we can remove it.
+    if (users.contains(name)) {
+      throw EnvoyException(fmt::format("basic auth: duplicate user '{}'", name));
```
Comment (zhaohuabing) [inline source/extensions/filters/http/basic_auth/config.cc:44]: I prefer to remove it. It's not so hard to tell which user is duplicated by looking at the config.

File: test/extensions/filters/http/basic_auth/filter_test.cc [hunk 4]
```diff
@@ -15,10 +15,9 @@ namespace BasicAuth {
...
 public:
   FilterTest() {
-    std::unique_ptr<absl::flat_hash_map<std::string, User>> users =
-        std::make_unique<absl::flat_hash_map<std::string, User>>();
-    users->insert({"user1", {"user1", "tESsBmE/yNY3lb6a0L6vVQEZNqw="}}); // user1:test1
-    users->insert({"user2", {"user2", "EJ9LPFDXsN9ynSmbxvjp75Bmlx8="}}); // user2:test2
+    UserMap users;
+    users.insert({"user1", {"user1", "tESsBmE/yNY3lb6a0L6vVQEZNqw="}}); // user1:test1
+    users.insert({"user2", {"user2", "EJ9LPFDXsN9ynSmbxvjp75Bmlx8="}}); // user2:test2
     config_ = std::make_unique<FilterConfig>(std::move(users), "stats", *stats_.rootScope());
```
Comment (zhaohuabing) [inline test/extensions/filters/http/basic_auth/filter_test.cc:21]: Haha, this is an interesting C++ hack to learn.

File: source/extensions/filters/http/basic_auth/basic_auth_filter.h [hunk 5]
```diff
@@ -36,15 +36,14 @@ struct User {
...
-    std::unique_ptr<const absl::flat_hash_map<std::string, User>>; // username, User
+using UserMap = absl::flat_hash_map<std::string, User>;
 
 /**
  * Configuration for the Basic Auth filter.
  */
 class FilterConfig {
 public:
-  FilterConfig(UserMapConstPtr users, const std::string& stats_prefix, Stats::Scope& scope);
+  FilterConfig(UserMap&& users, const std::string& stats_prefix, Stats::Scope& scope);
```
Comment (zhaohuabing) [inline source/extensions/filters/http/basic_auth/basic_auth_filter.h:46]: There is always an interesting C++ feature to learn :-)
What is the difference between this way and a const pointer?  When to choose what?

**Review by zhaohuabing — APPROVED** [review]
Nit: prefer not to output username, otherwise LGTM.

Thanks for improving it!

File: source/extensions/filters/http/basic_auth/config.cc (hunk 3 above)
Comment (wbpcode) [reply source/extensions/filters/http/basic_auth/config.cc:44]: got it.

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**